craft_application/launchpad/models/recipe.py

```diff
@@ -326,11 +326,16 @@ def build(
         return self._build(deadline, request_build_kwargs)
 
 
-class CharmRecipe(_StoreRecipe):
-    """A recipe for a charm.
+class _StandardRecipe(_StoreRecipe):
+    """A base class for recipes with common launchpad builds."""
 
-    https://api.launchpad.net/devel.html#charm_recipe
-    """
+    ARTIFACT: ClassVar[str]
+    """The artifact type this recipe creates: charm, rock, etc."""
+
+    @classmethod
+    def _lp_recipe(cls, lp: Launchpad) -> Any:  # noqa: ANN401 (use of any)
+        """Get the launchpad utility to manipulate recipes."""
+        raise NotImplementedError
```

Comment:
I think this should become an abstract method so we can get error messages sooner than runtime.

Reply:
I don't remember the specifics, but making abstract classes for these Launchpad models caused problems when I did the initial implementation.

Reply:
Interesting. Well if it can't be done then I guess this implementation is the best. @tigarmo if it doesn't work then let me know and I'll approve.

Reply:
Turns out that there's a specific order for `abstractmethod` and `classmethod` to work together. I added the decorator, but in practice I don't think it gives us much because these classes are never instantiated (all methods are classmethods), so the "abstract guard" doesn't come into play unfortunately.

But it still signals that the method needs to be overridden, so added. Thanks